### PR TITLE
allow method=nearest for fillna. see issue #9471.

### LINKED ISSUE #9471: Add method='nearest' to fillna
This should be a pretty straightforward followup on #9258 that will allow us to also support `align` with `method='nearest'`.

fillna internally relies on interpolate which already supports `method='nearest' so this *may* just work (it will needs lots of tests).

### PR DESCRIPTION
**Pull request by mgdadv**
closes #9471

### CODE REVIEW

File: pandas/core/generic.py [hunk 1]
```diff
@@ -2336,13 +2336,25 @@ def fillna(self, value=None, method=None, axis=None, inplace=False,
...
-            new_data = self._data.interpolate(method=method,
-                                              axis=axis,
-                                              limit=limit,
-                                              inplace=inplace,
-                                              coerce=True,
-                                              downcast=downcast)
+            method = com._clean_fill_method(method, allow_nearest=True)
+            new_data = None
+            if method == "nearest":
+                axis = 1
```
Comment (shoyer) [inline pandas/core/generic.py]: Why do you override the `axis` argument for nearest as a special case?

**Comment by shoyer** [comment]
This needs tests. Does it work on non-numeric dtypes? (e.g., strings?)

**Comment by mgdadv** [comment]
The latest commit should support all the methods from `interpolate()` and adds a test which checks for the same result as `interpolate()`. I'm not sure if this is a reasonable way to test this. The test is only for the DataFrame.

Unfortunately, it does not seem to work on non-numeric types, since scipy interpolation does not seem to support that case. I'm also unsure what distance metric should be used for strings.

Regarding your code comment, I changed the `axis=1` override to only happen if the incoming axis is None and one of the interpolate methods is used.

**Comment by mgdadv** [comment]
Regarding 'nearest', I can see these different scenarios:

1. index and values are both numeric.
   This should be mostly taken care of with calling out to `interpolate`.
2. index is numeric and values are non-numeric (in particular strings).
   This does not work with the above patch and should certainly be supported.
3. index is non-numeric and values are either numeric or non-numeric.
   The index still needs a distance metric (probably user defineable).
   This case seems more experimental and I'm not sure if it is even worthwhile doing.

Regarding case (2), this is what I am currently playing with:

```python
import numpy as np
import pandas as pd
from pandas.util.testing import assert_frame_equal

df = pd.DataFrame({'A': ["a",np.NaN,"c",np.NaN,"e"]}, index=[1,2,5,9,10])
expected = pd.DataFrame({'A': ["a","a","c","e","e"]}, index=[1,2,5,9,10])

nan_inds = df[df.isnull().any(axis=1)].index.values
df_no_nan = df.dropna()

result = df.copy()
result.loc[result.isnull().any(axis=1), 'A'] = [result.loc[df_no_nan.index[df_no_nan.index.get_loc(i, method='nearest')]].values[0] for i in nan_inds]

assert_frame_equal(result, expected)
```

This seems very convoluted. I would be grateful for hints.

**Comment by jreback** [comment]
closing pls reopen if/when updated

**Pull request closed without merging**